xapian-letor/ranker/ranker.cc

```diff
@@ -53,16 +341,147 @@ Ranker::labelcomparer(const FeatureVector & firstfv, const FeatureVector& second
+    train_model(list_fvecs);
+    save_model_to_metadata(model_key);
+}
+
+void
+Ranker::score(const string & query_file, const string & qrel_file,
+      const string & model_key, const string & output_file,
+      Xapian::doccount msetsize, const string & scorer_type, const Xapian::FeatureList & flist)
+{
+    // Set db
+    Xapian::Database letor_db(db_path);
+    // Load ranker model
+    load_model_from_metadata(model_key);
+    // Set scorer
+    Xapian::Internal::intrusive_ptr<Scorer> scorer;
+    if (scorer_type == "NDCGScore") {
+	scorer = new NDCGScore();
+    }
+    else {
+	throw LetorInternalError("Invalid Scorer type.");
+    }
+
+    Xapian::SimpleStopper mystopper(sw, sw + sizeof(sw) / sizeof(sw[0]));
+    Xapian::Stem stemmer("english");
+    Xapian::QueryParser parser;
+    parser.add_prefix("title", "S");
+    parser.add_prefix("subject", "S");
+    parser.set_database(Xapian::Database(letor_db));
+    parser.set_default_op(Xapian::Query::OP_OR);
+    parser.set_stemmer(stemmer);
+    parser.set_stemming_strategy(Xapian::QueryParser::STEM_SOME);
+    parser.set_stopper(&mystopper);
+
+    qrel = load_relevance(qrel_file);
+
+    string str1;
+    ifstream queryfile;
+    queryfile.open(query_file.c_str(), ios::in);
+    if(!queryfile.good()) {
+	throw Xapian::FileNotFoundError("No Query file found. Check path.");
+    }
+
+    ofstream out_file;
+    out_file.open(output_file);
+
+    double total_score = 0;
+    int num_queries = 0;
+    while (!queryfile.eof()) {
+	getline(queryfile, str1);
+	if (str1.empty()) {
+	    break;
+	}
+	++num_queries;
+
+	vector<string> token;
+	size_t j = 0;
+	while (j < str1.size()) {
+	    size_t i = str1.find_first_not_of(' ', j);
+	    if (i == string::npos) break;
+	    j = str1.find_first_of(' ', i);
+	    token.push_back(str1.substr(i, j - i));
+	}
+	// Query file is in the format: <qid> <query_string>
+	// Therefore, <qid> goes into token[0] and <query_string> to token[1]
+	// Exceptions for parse errors
+	if (token.size() != 2) {
+	    throw LetorParseError("Could not parse Query file at line:" + Xapian::Internal::str(num_queries));
+	}
+	string qid = token[0];
+	string querystr = token[1];
+	if (querystr.front() != '\'' || querystr.back() != '\'') {
+	    throw LetorParseError("Could not parse query string at line:" + Xapian::Internal::str(num_queries));
+	}
+	querystr.erase( 0, 1 ); // erase the first character (') from the front
+	querystr.erase( querystr.size() - 1); // erase the last character (')
+	if (querystr.empty()) {
+	    throw LetorParseError("Empty query string in query file at line:" + Xapian::Internal::str(num_queries));
+	}
+
+	string qq = querystr;
+	istringstream iss(querystr);
+	string title = "title:";
+	while (iss) {
+	    string t;
+	    iss >> t;
+	    if (t.empty())
+		break;
+	    string temp;
+	    temp.append(title);
+	    temp.append(t);
+	    temp.append(" ");
+	    temp.append(qq);
+	    qq = temp;
+	}
```

Comment:
That doesn't explain why there's a large piece of repeated code - the code to parse the query could be in a separate function called from each.